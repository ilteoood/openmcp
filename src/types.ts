import { OpenAPIV2, OpenAPIV3, type OpenAPIV3_1 } from "openapi-types";

export type PathItemObject =
	| OpenAPIV2.PathItemObject
	| OpenAPIV3.PathItemObject
	| OpenAPIV3_1.PathItemObject;

export type ParameterObject =
	| OpenAPIV2.ParameterObject
	| OpenAPIV3.ParameterObject
	| OpenAPIV3_1.ParameterObject;

export type ParametersWithRef = (
	| OpenAPIV2.PathItemObject
	| OpenAPIV3.PathItemObject
	| OpenAPIV3_1.PathItemObject
)["parameters"];

export const HTTPMethods = {
	...OpenAPIV2.HttpMethods,
	...OpenAPIV3.HttpMethods,
};
