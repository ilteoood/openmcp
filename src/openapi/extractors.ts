import type { OpenAPIV3 } from "openapi-types";
import {
	createForObject,
	createFromParameter,
	createFromParameters,
	createFromSchema,
	oneOf,
} from "../schema.js";
import type {
	ExtractedPath,
	FormDataSchema,
	ParamRequestObject,
	ParameterObject,
	ParametersWithRef,
	PathItemObject,
	RequestBodySchema,
} from "../types.js";

const extractNameFromPath = (path: string) => {
	return path
		.split("/")
		.filter((segment) => segment && !segment.startsWith("{"))
		.join("-");
};

export const basePathItemExtractor = (
	path: string,
	pathItem: PathItemObject,
	method: OpenAPIV3.HttpMethods,
): ExtractedPath => {
	const pathItemMethod = (pathItem as OpenAPIV3.PathItemObject)[method];
	return {
		name: pathItemMethod?.operationId ?? extractNameFromPath(path),
		method,
		path,
		parameters: extractParameters(
			pathItem.parameters,
			pathItemMethod?.parameters,
		),
		request: createForObject(),
		security: oneOf(
			pathItemMethod?.security?.flatMap((security) => Object.keys(security)),
		),
	};
};

export const pathItemWithBodyExtractor = (
	path: string,
	pathItem: PathItemObject,
	method: OpenAPIV3.HttpMethods,
) => {
	return {
		...basePathItemExtractor(path, pathItem, method),
		request: requestExtractor(pathItem),
	};
};

const SUPPORTED_PARAMETERS = new Set(["path", "query", "header", "cookie"]);

export const extractParameters = (
	pathParameters?: ParametersWithRef,
	methodParameters?: ParametersWithRef,
) => {
	const parameters = (
		(pathParameters ?? []).concat(methodParameters ?? []) as ParameterObject[]
	).filter((parameter) => SUPPORTED_PARAMETERS.has(parameter.in));

	const reducedParameters = parameters.reduce(
		(acc, param) => {
			const paramSource = param.in as keyof typeof acc;
			const paramDestination = acc[paramSource];
			paramDestination[param.name] = createFromParameter(param);
			return acc;
		},
		{
			path: {} as ParamRequestObject,
			query: {} as ParamRequestObject,
			header: {} as ParamRequestObject,
			cookie: {} as ParamRequestObject,
		},
	);

	return createForObject({
		path: createForObject(reducedParameters.path),
		query: createForObject(reducedParameters.query),
		header: createForObject(reducedParameters.header),
		cookie: createForObject(reducedParameters.cookie),
	});
};

export const requestExtractor = (pathItem: PathItemObject) =>
	createForObject({
		body: extractRequestBody(pathItem),
		formData: extractFormData(pathItem),
	});

const extractFormData = (pathItem: PathItemObject): FormDataSchema => {
	const formDataParameters = (
		pathItem.post?.parameters as ParameterObject[] | undefined
	)?.filter((param) => param.in === "formData");

	let requestFormData: FormDataSchema;

	if (formDataParameters) {
		requestFormData = createFromParameters(formDataParameters);
	} else {
		const formDataBody = (
			(pathItem.post as OpenAPIV3.OperationObject)
				?.requestBody as OpenAPIV3.RequestBodyObject
		)?.content?.["application/x-www-form-urlencoded"];
		requestFormData = formDataBody
			? createFromSchema(formDataBody.schema)
			: createForObject();
	}

	return requestFormData;
};

const extractRequestBody = (pathItem: PathItemObject): RequestBodySchema => {
	const bodyParameters = (
		pathItem.post?.parameters as ParameterObject[] | undefined
	)?.filter((param) => param.in === "body");

	let requestBody: RequestBodySchema;

	if (bodyParameters) {
		requestBody = createFromParameters(bodyParameters);
	} else {
		const jsonBody = (
			(pathItem.post as OpenAPIV3.OperationObject)
				?.requestBody as OpenAPIV3.RequestBodyObject
		)?.content?.["application/json"];
		requestBody = jsonBody
			? createFromSchema(jsonBody.schema)
			: createForObject();
	}

	return requestBody;
};
