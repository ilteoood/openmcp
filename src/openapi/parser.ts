import SwaggerParser from "@apidevtools/swagger-parser";
import type { OpenAPI } from "openapi-types";
import {
	HTTPMethods,
	type PathItemObject
} from "../types.js";
import { basePathItemExtractor, requestExtractor } from "./extractors.js";

export class OpenAPIParser {
	private constructor(private swaggerDocument: OpenAPI.Document) {}

	static async from(resourceLocator: string): Promise<OpenAPIParser> {
		const parsedDocument = await SwaggerParser.parse(resourceLocator);
		const dereferencedDocument =
			await SwaggerParser.dereference(parsedDocument);
		return new OpenAPIParser(dereferencedDocument);
	}

	public getPaths() {
		const pathsDefinition = this.swaggerDocument.paths || {};
		const paths = Object.keys(pathsDefinition);
		const pathDefinition = [];

		for (const path of paths) {
			const pathItem = pathsDefinition[path];

			if (!pathItem) continue;

			if (pathItem.get) {
				pathDefinition.push(this.handleGet(pathItem));
			}
			if (pathItem.post) {
				pathDefinition.push(this.handlePost(pathItem));
			}
			if (pathItem.put) {
				pathDefinition.push(this.handlePut(pathItem));
			}
			if (pathItem.delete) {
				pathDefinition.push(this.handleDelete(pathItem));
			}
			if (pathItem.patch) {
				pathDefinition.push(this.handlePatch(pathItem));
			}
			if (pathItem.options) {
				pathDefinition.push(this.handleOptions(pathItem));
			}
			if (pathItem.head) {
				pathDefinition.push(this.handleHead(pathItem));
			}
			if ("trace" in pathItem) {
				pathDefinition.push(this.handleTrace(pathItem));
			}
		}

		return pathDefinition;
	}

	public getInfo() {
		const info = this.swaggerDocument.info;
		return {
			name: info?.title,
			version: info?.version,
		}
	}

	private handleGet(pathItem: PathItemObject) {
		return basePathItemExtractor(pathItem, HTTPMethods.GET);
	}

	private handlePost(pathItem: PathItemObject) {
		return {
			...basePathItemExtractor(pathItem, HTTPMethods.POST),
			request: requestExtractor(pathItem),
		};
	}

	private handlePut(pathItem: PathItemObject) {
		return {
			...basePathItemExtractor(pathItem, HTTPMethods.PUT),
			request: requestExtractor(pathItem),
		};
	}

	private handleDelete(pathItem: PathItemObject) {
		return basePathItemExtractor(pathItem, HTTPMethods.DELETE);
	}

	private handlePatch(pathItem: PathItemObject) {
		return {
			...basePathItemExtractor(pathItem, HTTPMethods.PATCH),
			request: requestExtractor(pathItem),
		};
	}

	private handleOptions(pathItem: PathItemObject) {
		return basePathItemExtractor(pathItem, HTTPMethods.OPTIONS);
	}

	private handleHead(pathItem: PathItemObject) {
		return basePathItemExtractor(pathItem, HTTPMethods.HEAD);
	}

	private handleTrace(pathItem: PathItemObject) {
		return basePathItemExtractor(pathItem, HTTPMethods.TRACE);
	}
}
