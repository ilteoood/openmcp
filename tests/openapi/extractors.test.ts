import { describe, expect, it } from "vitest";
import { basePathItemExtractor } from "../../src/openapi/extractors.js";
import { OpenAPIParser } from "../../src/openapi/parser.js";
import { HTTPMethods } from "../../src/types.js";
import { buildFixturePath, loadFixtures } from "../utils.js";

describe("parser", () => {
	it("can correctly return the informations", async () => {
		const schemaPath = buildFixturePath("schema.json");

		const openApiParser = await OpenAPIParser.from(schemaPath);

		expect(openApiParser.getInfo()).toStrictEqual({
			name: "Swagger Petstore - OpenAPI 3.0",
			version: "1.0.26",
		});
	});

	it("can correctly return base path item", async () => {
		const openApiDocument = await loadFixtures("schema.json");

		const result = await basePathItemExtractor(
			"/pet/{petId}",
			openApiDocument.paths["/pet/{petId}"],
			HTTPMethods.POST,
		);

		expect(result).toMatchSnapshot();
	});

	it("can correctly return base path item - generated name", async () => {
		const openApiDocument = await loadFixtures("schema.json");

		const result = await basePathItemExtractor(
			"/pets/location",
			openApiDocument.paths["/pets/location"],
			HTTPMethods.POST,
		);

		expect(result).toMatchSnapshot();
	});
});
