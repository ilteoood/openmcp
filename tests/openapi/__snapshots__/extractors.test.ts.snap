// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extractors > should create a zod schema for a 'array' type 1`] = `
ZodOptional {
  "_def": {
    "description": "List of owners of the pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodArray {
      "_def": {
        "description": "List of owners of the pet that needs to be updated",
        "exactLength": null,
        "maxLength": null,
        "minLength": null,
        "type": ZodAny {
          "_any": true,
          "_def": {
            "typeName": "ZodAny",
          },
          "and": [Function],
          "array": [Function],
          "brand": [Function],
          "catch": [Function],
          "default": [Function],
          "describe": [Function],
          "isNullable": [Function],
          "isOptional": [Function],
          "nullable": [Function],
          "nullish": [Function],
          "optional": [Function],
          "or": [Function],
          "parse": [Function],
          "parseAsync": [Function],
          "pipe": [Function],
          "promise": [Function],
          "readonly": [Function],
          "refine": [Function],
          "refinement": [Function],
          "safeParse": [Function],
          "safeParseAsync": [Function],
          "spa": [Function],
          "superRefine": [Function],
          "transform": [Function],
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        "typeName": "ZodArray",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`extractors > should create a zod schema for a 'boolean' type 1`] = `
ZodOptional {
  "_def": {
    "description": "Status of pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodBoolean {
      "_def": {
        "coerce": false,
        "description": "Status of pet that needs to be updated",
        "typeName": "ZodBoolean",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`extractors > should create a zod schema for a 'integer not required' type 1`] = `
ZodOptional {
  "_def": {
    "description": "Age of the pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodNumber {
      "_def": {
        "checks": [
          {
            "kind": "int",
            "message": undefined,
          },
        ],
        "coerce": false,
        "description": "Age of the pet that needs to be updated",
        "typeName": "ZodNumber",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "max": [Function],
      "min": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "step": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`extractors > should create a zod schema for a 'integer' type 1`] = `
ZodNumber {
  "_def": {
    "checks": [
      {
        "kind": "int",
        "message": undefined,
      },
    ],
    "coerce": false,
    "description": "ID of pet that needs to be updated",
    "typeName": "ZodNumber",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "max": [Function],
  "min": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "step": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`extractors > should create a zod schema for a 'number' type 1`] = `
ZodOptional {
  "_def": {
    "description": "Height of the pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodNumber {
      "_def": {
        "checks": [],
        "coerce": false,
        "description": "Height of the pet that needs to be updated",
        "typeName": "ZodNumber",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "max": [Function],
      "min": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "step": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`extractors > should create a zod schema for a 'object' type 1`] = `
ZodOptional {
  "_def": {
    "description": "General informations about the pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodObject {
      "_cached": null,
      "_def": {
        "catchall": ZodNever {
          "_def": {
            "typeName": "ZodNever",
          },
          "and": [Function],
          "array": [Function],
          "brand": [Function],
          "catch": [Function],
          "default": [Function],
          "describe": [Function],
          "isNullable": [Function],
          "isOptional": [Function],
          "nullable": [Function],
          "nullish": [Function],
          "optional": [Function],
          "or": [Function],
          "parse": [Function],
          "parseAsync": [Function],
          "pipe": [Function],
          "promise": [Function],
          "readonly": [Function],
          "refine": [Function],
          "refinement": [Function],
          "safeParse": [Function],
          "safeParseAsync": [Function],
          "spa": [Function],
          "superRefine": [Function],
          "transform": [Function],
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        "description": "General informations about the pet that needs to be updated",
        "shape": [Function],
        "typeName": "ZodObject",
        "unknownKeys": "strip",
      },
      "and": [Function],
      "array": [Function],
      "augment": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nonstrict": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`extractors > should create a zod schema for a 'string' type 1`] = `
ZodOptional {
  "_def": {
    "description": "Name of pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodString {
      "_def": {
        "checks": [],
        "coerce": false,
        "description": "Name of pet that needs to be updated",
        "typeName": "ZodString",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;
